design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+   * @return column value.
+   */
+  String getString(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as an integer.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  Integer getInt(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as an integer.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Integer getInt(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a long.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  Long getLong(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a long.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Long getLong(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a double.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  Double getDouble(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a double.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Double getDouble(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a boolean.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  Boolean getBoolean(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a boolean.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Boolean getBoolean(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a decimal.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  BigDecimal getDecimal(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a decimal.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  BigDecimal getDecimal(String columnName);
+}
+```
+
+We considered representing column types in a more structured form (rather than plain strings) to accomodate complex/nested data types but felt this added complexity would not be helpful for most use cases. (This would also require a server-side change, so it's a fair bit of additional work.)
+
+For `getDecimal(...)` in the `Row` interface, rather than trying to parse the column type to extract the precision and scale, we will simply convert the value to a BigDecimal without explicitly specifying the precision and scale. We could also add an option for users to specify the scale and precision in the getter, if we think that would be useful.
+
+We could also add `getList(...)` and `getMap(...)` methods to the `Row` interface, but it's not clear to me how valuable this would be given that the nested data types would not be known.
```

Comment:
Why can't we expose the nested data types?

Reply:
Agreed, JsonObject/JsonArray